lib/view_component/compiler.rb

```diff
@@ -103,266 +50,273 @@ def renders_template_for?(variant, format)
+          .map do |normalized_variant_name|
+          variant_pairs
+            .select { |variant_pair| variant_pair.last == normalized_variant_name }
+            .map { |variant_pair| variant_pair.first }
+        end
 
-          unless colliding_variants.empty?
-            errors <<
-              "Colliding templates #{colliding_variants.sort.map { |v| "'#{v}'" }.to_sentence} " \
-              "found in #{component_class}."
-          end
+      colliding_normalized_variants.each do |variants|
+        errors << "Colliding templates #{variants.sort.map { |v| "'#{v}'" }.to_sentence} found in #{@component}."
+      end
 
-          errors
-        end
+      raise TemplateError.new(errors) if errors.any? && raise_errors
+
+      errors
     end
 
-    def templates
+    def gather_templates
       @templates ||=
         begin
-          extensions = ActionView::Template.template_handler_extensions
-
-          component_class.sidecar_files(extensions).each_with_object([]) do |path, memo|
+          templates = @component.sidecar_files(
+            ActionView::Template.template_handler_extensions
+          ).map do |path|
             pieces = File.basename(path).split(".")
-            memo << {
+
+            out = Template.new(
+              component: @component,
+              type: :file,
               path: path,
-              format: pieces[1..-2].join(".").split("+").first&.to_sym,
-              variant: pieces[1..-2].join(".").split("+").second&.to_sym,
-              handler: pieces.last
-            }
+              lineno: 0,
+              extension: pieces.last,
+              this_format: pieces[1..-2].join(".").split("+").first&.to_sym,
+              variant: pieces[1..-2].join(".").split("+").second&.to_sym
+            )
+
+            @rendered_templates << [out.variant, out.this_format]
+
+            out
           end
-        end
-    end
 
-    def inline_calls
-      @inline_calls ||=
-        begin
-          # Fetch only ViewComponent ancestor classes to limit the scope of
-          # finding inline calls
-          view_component_ancestors =
-            (
-              component_class.ancestors.take_while { |ancestor| ancestor != ViewComponent::Base } -
-              component_class.included_modules
+          (
+            @component.ancestors.take_while { |ancestor| ancestor != ViewComponent::Base } - @component.included_modules
+          ).flat_map { |ancestor| ancestor.instance_methods(false).grep(/^call(_|$)/) }
+            .uniq
+            .each do |method_name|
+              templates << Template.new(
+                component: @component,
+                type: :inline_call,
+                this_format: ViewComponent::Base::DEFAULT_FORMAT,
+                variant: method_name.to_s.include?("call_") ? method_name.to_s.sub("call_", "").to_sym : nil,
+                method_name: method_name,
+                defined_on_self: @component.instance_methods(false).include?(method_name)
```

Comment:
How often do we call`@component.instance_methods(false)`? Just wondering if that's something we should memoize in some way (e.g. store in a local variable) so we can avoid extra processing/allocations

Reply:
Agreed!